msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java

```diff
@@ -1143,10 +1144,26 @@ public static GenerateShrCommandParameters createGenerateShrCommandParameters(
     public static List<Map.Entry<String, String>> appendToExtraQueryParametersIfWebAuthnCapable(
             @Nullable final List<Map.Entry<String, String>> queryStringParameters,
             @NonNull final PublicClientApplicationConfiguration configuration) {
-        if (configuration.isWebauthnCapable()) {
-            final Map.Entry<String, String> webauthnExtraParameter = new AbstractMap.SimpleEntry<>(
-                    FidoConstants.WEBAUTHN_QUERY_PARAMETER_FIELD,
-                    FidoConstants.WEBAUTHN_QUERY_PARAMETER_VALUE);
+        final String methodTag = TAG + ":appendToExtraQueryParametersIfWebAuthnCapable";
+        final Map.Entry<String, String> webauthnExtraParameter = new AbstractMap.SimpleEntry<>(
+                FidoConstants.WEBAUTHN_QUERY_PARAMETER_FIELD,
+                FidoConstants.WEBAUTHN_QUERY_PARAMETER_VALUE);
+        // Check the OS version. As of the time this is written, passkeys are only supported on devices that run Android 9 (API 28) or higher.
+        // https://developer.android.com/identity/sign-in/credential-manager
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
+            Logger.info(methodTag, "Device is running on an Android version less than 9 (API 28), which is the minimum level for passkeys.");
+            // If we don't want to add this query string param, then we should also remove other instances of it that might be already present from MSAL/OneAuth-MSAL.
+            if (queryStringParameters != null) {
+                try {
+                    queryStringParameters.remove(webauthnExtraParameter);
+                } catch (final UnsupportedOperationException e) {
+                    final List<Map.Entry<String, String>> result = new ArrayList<>(queryStringParameters);
+                    result.remove(webauthnExtraParameter);
+                    return result;
+                }
+            }
+        }
```

Comment:
This is basically us "correcting" the apps?

Reply:
Also I'm wondering if this whole OS check and removal of qp can go into a helper method in common? and then invoke that method in both MSAL and Broker?

Reply:
Addressed this comment by creating the FidoUtil class in common: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2450/commits/6857ba4522f53a10538c82123021102d199b72ea
And this commit for MSAL: 818ff76d2ace0eb3b3040a7d48ea07024fa60e77